pkg/optimizer/ebpf/conn/conn.go

```diff
@@ -0,0 +1,357 @@
+
+    if (container_id_filter()) {
+        return 0;
+    };
+
+    bpf_probe_read(&event->position, sizeof(event->position), pos);
+    fill_file_path(file, event->path);
+    vfs_read_start_trace.update(&pid, event);
+
+    return 0;
+}
+
+// vfs_read will not nested , this is safe
+int trace_read_return(struct pt_regs *ctx)
+{
+    int ret = PT_REGS_RC(ctx);
+    u64 pid = bpf_get_current_pid_tgid();
+    struct request_info *event = vfs_read_start_trace.lookup(&pid);
+    if (!event)
+        return 0;
+    if (ret <= 0)
+        return 0;
+
+    event->length = ret;
+    fille_access_events.perf_submit(ctx, event, sizeof(struct request_info));
+
+    return 0;
+}
+
+int trace_vfs_read_entry(struct pt_regs *ctx, struct file *file, char __user *buf, size_t count, loff_t *pos)
+{
+    return trace_read_entry(ctx, file, pos);
+}
+
+int trace_splice_read_entry(struct pt_regs *ctx, struct file *in, loff_t *ppos,
+				 struct pipe_inode_info *pipe, size_t len,
+				 unsigned int flags)
+{
+    return trace_read_entry(ctx, in, ppos);
+}
+
+int trace_vfs_readv_entry(struct pt_regs *ctx, struct file *file, const struct iovec __user *vec,
+		  unsigned long vlen, loff_t *pos, rwf_t flags)
+{
+    return trace_read_entry(ctx, file, pos);
+}
+
+//readahead may influence the page fault collection data, disable readahead for tracing mmap read.
+int trace_page_fault(struct pt_regs *ctx, struct vm_fault *vmf)
+{
+    int zero = 0;
+    u64 page_size = 4096;
+    u64 pid = bpf_get_current_pid_tgid();
+    struct request_info *event = event_buf.lookup(&zero);
+    struct vm_area_struct *vma = vmf->vma;
+    struct file *file = vma->vm_file;
+    int container_id_len;
+
+    if (!vma)
+        return 0;
+
+    // only trace file backed page fault
+    if (!file)
+        return 0;
+
+    if (!event)
+        return 0;
+
+    if (file_filter(file)) {
+        return 0;
+    }
+
+    //skip page fault for write
+    if ((vmf->flags & FAULT_FLAG_WRITE) && (vma->vm_flags & VM_SHARED))
+        return 0;
+
+    bpf_get_current_comm(event->comm, sizeof(event->comm));
+    
+    if (container_id_filter()) {
+        return 0;
+    };
+
+    fill_file_path(file, event->path);
+    event->position = (vma->vm_pgoff + vmf->pgoff) * page_size;
+    event->length = page_size;
+    fille_access_events.perf_submit(ctx, event, sizeof(struct request_info));
+
+    return 0;
+}
+`
+
+func kprobeSyscall(m *bpf.Module, syscall string, kprobeEntry string) error {
+	vfsReadKprobe, err := m.LoadKprobe(kprobeEntry)
+	if err != nil {
+		return errors.Wrapf(err, "load entry %s", kprobeEntry)
+	}
+
+	err = m.AttachKprobe(syscall, vfsReadKprobe, -1)
+	if err != nil {
+		return errors.Wrapf(err, "attach entry %s", kprobeEntry)
+	}
+	return nil
+}
+
+func InitKprobeTable(id string) (*bpf.Module, *bpf.Table, error) {
+	m := bpf.NewModule(sourceCode, []string{})
+
+	if err := kprobeSyscall(m, "vfs_read", "trace_vfs_read_entry"); err != nil {
```

Comment:
if we are filtering overlayfs file operations, why not just hook on `ovl_read_iter`?